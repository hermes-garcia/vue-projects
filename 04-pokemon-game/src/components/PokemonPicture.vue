<template>
  <h3>Picture</h3>
</template>

<script>
export default {

}
</script>

<style>

</style>