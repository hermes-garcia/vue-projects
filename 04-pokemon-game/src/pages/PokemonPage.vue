<template>

  <div v-if="selectedPokemon">
    <h1>Who is that Pokemon?</h1>

    <PokemonPicture
        :pokemonId="selectedPokemon.id"
        :showPokemon="showPokemon"
    />

    <PokemonOptions
        :pokemon="pokemonArr"
        @selection="checkAnswer"
    />

    <div class="answer" v-if="showAnswer">
      <h2 class="fade-in">{{message}}</h2>
      <button @click="newGame">New Game</button>
    </div>
  </div>

  <h1 v-else>Fetching pokemon...</h1>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture';
import PokemonOptions from '@/components/PokemonOptions';
import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
  components: {PokemonOptions, PokemonPicture},
  data() {
    return {
      pokemonArr: [],
      selectedPokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      const rndInt = Math.floor(Math.random() * 4);
      this.selectedPokemon = this.pokemonArr[rndInt];
    },
    checkAnswer(selectedId) {
      this.showPokemon = true;
      this.showAnswer = true;

      if (selectedId === this.selectedPokemon.id) {
        this.message = `Correct! It's ${this.selectedPokemon.name}`
      } else {
        this.message = `Oops! It was ${this.selectedPokemon.name}`
      }
    },
    newGame() {
      this.pokemonArr = [];
      this.selectedPokemon = null;
      this.showPokemon = false;
      this.showAnswer = false;
      this.message = '';
      this.mixPokemonArray();
    }
  },
  mounted() {
    this.mixPokemonArray();
  }
}
</script>

<style scoped>
.answer {
  display: flex;
  justify-content: center;
  flex-direction: column;

  h2 {
  }

  button {
    margin: 0 auto;
  }
}
</style>