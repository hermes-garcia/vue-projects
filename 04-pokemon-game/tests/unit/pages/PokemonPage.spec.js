import {mount, shallowMount} from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import {pokemonListMock} from '../../mocks/pokemonList.mock';
import PokemonPicture from '@/components/PokemonPicture';
import PokemonOptions from '@/components/PokemonOptions';

describe('PokemonPage component', function() {
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('should call mixPokemonArray on mount', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    });

    test('should match snapshot after load pokemon', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonListMock,
                    selectedPokemon: pokemonListMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    });

    test('should show PokemonPicture and PokemonOptions components ', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonListMock,
                    selectedPokemon: pokemonListMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        })
        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')
        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()
        expect(picture.attributes('pokemonid')).toBe('5')
        expect(picture.attributes('showpokemon')).toBe('false')
        expect(options.attributes('pokemon')).toBeTruthy()
    });

    test('checkAnswer testing', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonListMock,
                    selectedPokemon: pokemonListMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        })

        await wrapper.vm.checkAnswer(pokemonListMock[0].id);

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correct! It's ${pokemonListMock[0].name}`)
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.vm.showAnswer).toBe(true)

        await wrapper.vm.checkAnswer(pokemonListMock[1].id);

        expect(wrapper.find('h2').text()).toBe(`Oops! It was ${pokemonListMock[0].name}`)
    });
});