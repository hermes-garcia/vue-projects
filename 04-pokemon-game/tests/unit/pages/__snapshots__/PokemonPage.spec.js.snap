// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component should match snapshot after load pokemon 1`] = `
<div>
  <h1>Who is that Pokemon?</h1>
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemon="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
